engines/twp/actorlib.cpp

```diff
@@ -0,0 +1,1113 @@
+	Object::say(obj, texts, obj->_talkColor);
+	return 0;
+}
+
+// Causes an actor to say a line of dialog and play the appropriate talking animations.
+// In the first example, the actor ray will say the line.
+// In the second, the selected actor will say the line.
+// In the third example, the first line is displayed, then the second one.
+// See also:
+// - `mumbleLine method`
+static SQInteger sayLine(HSQUIRRELVM v) {
+	g_engine->stopTalking();
+	return sayOrMumbleLine(v);
+}
+
+// Say a line of dialog and play the appropriate talking animations.
+// In the first example, the actor ray will say the line.
+// In the second, the selected actor will say the line.
+// In the third example, the first line is displayed, then the second one.
+// See also:
+// - `mumbleLine method`
+static SQInteger sayLineAt(HSQUIRRELVM v) {
+	int x, y;
+	Common::String text;
+	float duration = -1.0f;
+	if (SQ_FAILED(sqget(v, 2, x)))
+		return sq_throwerror(v, "failed to get x");
+	if (SQ_FAILED(sqget(v, 3, y)))
+		return sq_throwerror(v, "failed to get y");
+	Color color;
+	if (sq_gettype(v, 4) == OT_INTEGER) {
+		int c;
+		if (SQ_FAILED(sqget(v, 4, c)))
+			return sq_throwerror(v, "failed to get color");
+		color = Color::rgb(c);
+		if (SQ_FAILED(sqget(v, 5, duration)))
+			return sq_throwerror(v, "failed to get duration");
+		if (SQ_FAILED(sqget(v, 6, text)))
+			return sq_throwerror(v, "failed to get text");
+	} else {
+		Common::SharedPtr<Object> actor = sqactor(v, 4);
+		if (!actor)
+			return sq_throwerror(v, "failed to get actor");
+		Math::Vector2d pos = g_engine->roomToScreen(actor->_node->getAbsPos());
+		x = pos.getX();
+		y = pos.getY();
+		color = actor->_talkColor;
+		if (SQ_FAILED(sqget(v, 6, text)))
+			return sq_throwerror(v, "failed to get text");
+	}
+
+	warning("TODO: saylineAt: (%d,%d) text=%s color=%s duration=%f", x, y, text.c_str(), color.toStr().c_str(), duration);
+	return 0;
+}
+
+// returns true if the specified actor is currently in the screen.
+static SQInteger isActorOnScreen(HSQUIRRELVM v) {
+	Common::SharedPtr<Object> obj = sqobj(v, 2);
+	if (!obj)
+		return sq_throwerror(v, "failed to get actor/object");
+
+	if (obj->_room != g_engine->_room) {
+		sqpush(v, false);
+	} else {
+		Math::Vector2d pos = obj->_node->getPos() - g_engine->getGfx().cameraPos();
+		Math::Vector2d size = g_engine->getGfx().camera();
+		bool isOnScreen = Common::Rect(0.0f, 0.0f, size.getX(), size.getY()).contains(pos.getX(), pos.getY());
+		sqpush(v, isOnScreen);
+	}
+	return 1;
+}
+
+static SQInteger isActorSelectable(HSQUIRRELVM v) {
+	Common::SharedPtr<Object> actor = sqactor(v, 2);
+	if (!actor)
+		return sq_throwerror(v, "failed to get actor");
+	ActorSlot *slot = g_engine->_hud.actorSlot(actor);
+	bool selectable = slot && slot->selectable;
+	sqpush(v, selectable);
+	return 1;
+}
+
+// If an actor is specified, returns true otherwise returns false.
+static SQInteger is_actor(HSQUIRRELVM v) {
+	Common::SharedPtr<Object> actor = sqactor(v, 2);
```

Comment:
I wonder, what is the purpose of declaring it as `SharedPtr`, when it goes out of scope right away?

Reply:
The method is returning a `SharedPtr`, do I have the choice? Maybe I'm missing something?